awx/main/scheduler/task_manager.py

```diff
@@ -394,8 +396,18 @@ def gen_dep_for_job(self, task):
             if not inventory_source.update_on_launch:
                 continue
             latest_inventory_update = self.get_latest_inventory_update(inventory_source)
-            if self.should_update_inventory_source(task, latest_inventory_update):
-                inventory_task = self.create_inventory_update(task, inventory_source)
+            # necessary for branch override to work properly
+            # if inventory source uses the same project as task, and the last inventory update has a different scm_branch, we should set branch override
+            if (
+                inventory_source.source == "scm"
+                and inventory_source.source_project_id == task.project_id
+                and latest_inventory_update.scm_branch != task.scm_branch
+            ):
+                should_set_scm_branch = True
+            else:
+                should_set_scm_branch = False
+            if should_set_scm_branch or self.should_update_inventory_source(task, latest_inventory_update):
```

Comment:
> 2. Inventory Source must have Update on Launch enabled

Doesn't this criteria conflict with the requirement? I see `should_set_scm_branch` has no bearing on the InventorySource update-on-launch setting, and the `or` means that it being true will spawn an update, spawning updates even when update-on-launch is False, which would be bad.

I liked what you were doing before where you tested this condition in `should_update_inventory_source`, which could more naturally come after the update-on-launch check.

Reply:
yeah i didn't want to repeat the logic for checking whether we should set scm_branch twice

once in `should_update_inventory_source` and another time in `create_inventory_update`

this was my way of trying to consolidate it by moving it outside of should_update method

also if `update_on_launch` is not enabled, we hit a continue above so we don't even hit this logic. So in that way it is an implicit requirement, but for code readability this is kind of messy

Reply:
As I'm internally churning through this in my mind, I set aside a proposal to rename `should_update_inventory_source` to `inventory_update_passed_cache_timeout`. I can see that the method checks `if latest_inventory_update.inventory_source.update_on_launch is True`, which seems to be the 3rd time that check is done. To begin with `self.all_inventory_sources` is filtered to update-on-launch sources. Then we check here, and then again in that method.

I'm still trying to catch up... so onto my next question.

Let's say that the job has an `scm_branch` which is non-blank. In what cases should `create_inventory_update` _not_ set the `scm_branch`? Implication of my question is that it should always do that. With your enhancement here, it _seems_ like every inventory update dependency will have the same `scm_branch` as the job that triggered it. If this is always true, then `should_set_scm_branch` shouldn't be a parameter, because the only time it would be `False` would be when job's `scm_branch` is blank, in which case it doesn't matter if you do or don't set it.

Reply:
answering my own question - it should not inherit the `scm_branch` if the source is not "scm". But if it is scm source, then it should always get the scm_branch from the job, unless there's a corner case I'm missing.